substrate/frame/collective/src/lib.rs

```diff
@@ -666,6 +677,41 @@ fn get_result_weight(result: DispatchResultWithPostInfo) -> Option<Weight> {
 }
 
 impl<T: Config<I>, I: 'static> Pallet<T, I> {
+	/// The active proposals.
+	#[cfg(any(feature = "std", feature = "runtime-benchmarks"))]
+	pub fn proposals() -> BoundedVec<T::Hash, T::MaxProposals> {
+		Voting::<T, I>::iter_keys()
+			.collect::<Vec<_>>()
+			.try_into()
+			.expect("proposals should always be less than MaxProposals; qed")
+	}
+
+	/// The bounded counterpart of an active proposal.
+	pub fn proposal_of(proposal_hash: &T::Hash) -> Option<<T as Config<I>>::Proposal> {
+		match Voting::<T, I>::contains_key(proposal_hash) {
+			false => None,
+			true => {
+				let Ok(encoded) = T::Preimages::fetch(proposal_hash, None) else {
+					log::error!(
+						target: LOG_TARGET,
+						"Failed to fetch preimage for proposal hash {:?} contained in Voting.",
+						proposal_hash,
+					);
+					return None
+				};
+				let Ok(proposal) = <T as Config<I>>::Proposal::decode(&mut &encoded[..]) else {
+					log::error!(
+						target: LOG_TARGET,
+						"Failed to decode stored Proposal with preimage hash {:?}",
+						proposal_hash,
+					);
+					return None
+				};
+				Some(proposal)
+			},
+		}
+	}
```

Comment:
This code is difficult to follow. 

Always use early return and avoid nesting at all costs.

Example of a the same thing with less nesting (NOT tested):

```rust
pub fn proposal_of(proposal_hash: &T::Hash) -> Result<Option<<T as Config<I>>::Proposal>, SomeErrorType> {
    if !Voting::<T, I>::contains_key(proposal_hash) {
        return Ok(None);
    }

    let encoded = T::Preimages::fetch(proposal_hash, None).map_err(|_| {
        log::error!(
            target: LOG_TARGET,
            "Failed to fetch preimage for proposal hash {:?} contained in Voting.",
            proposal_hash,
        );
        SomeErrorType::FetchingPreimageError
    })?;

    let proposal = <T as Config<I>>::Proposal::decode(&mut &encoded[..]).map_err(|_| {
        log::error!(
            target: LOG_TARGET,
            "Failed to decode stored Proposal with preimage hash {:?}",
            proposal_hash,
        );
        SomeErrorType::DecodingError
    })?;

    Ok(Some(proposal))
}
```

Reply:
Hmm, gotcha. If that's the case we can leave the function signature, but let's still flatten the indentation a bit :)

Reply:
I have already done it in the latest commit!